pkg/analysis/passes/jargon/jargon_test.go

```diff
@@ -0,0 +1,65 @@
+package jargon
+
+import (
+	"testing"
+
+	"github.com/grafana/plugin-validator/pkg/analysis"
+	"github.com/grafana/plugin-validator/pkg/analysis/passes/readme"
+	"github.com/stretchr/testify/assert"
+)
+
+var cleanReadme = []byte(`
+# Test
+
+This is a README without jargon in it.
+`)
+
+var jargonReadme = []byte(`
+# Test
+
+This is a README with jargon on it
+
+# Development instructions
+
+yarn install
+
+`)
+
+func TestClean(t *testing.T) {
+	var invoked bool
+
+	pass := &analysis.Pass{
+		ResultOf: map[*analysis.Analyzer]interface{}{
+			readme.Analyzer: cleanReadme,
+			Analyzer:        nil,
+		},
+		Report: func(_ string, _ analysis.Diagnostic) {
+			invoked = true
+		},
+	}
+
+	_, err := Analyzer.Run(pass)
+	assert.NoError(t, err)
+
+	// should not call the report function
+	assert.False(t, invoked)
+
+}
+
+func TestWithJargon(t *testing.T) {
+
+	pass := &analysis.Pass{
+		ResultOf: map[*analysis.Analyzer]interface{}{
+			readme.Analyzer: jargonReadme,
+			Analyzer:        nil,
+		},
+		Report: func(_ string, d analysis.Diagnostic) {
+			assert.Equal(t, "README.md contains developer jargon: (yarn)", d.Title)
+		},
+	}
+
+	_, err := Analyzer.Run(pass)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
```

Comment:
We also need to make sure that the `Report` anonymous function runs (which also contains the assertion), otherwise the test will pass if `Report` is not called (and the assertion about "contains yarn" won't be evaluated either):

```suggestion
func TestWithJargon(t *testing.T) {
	var invoked bool

	pass := &analysis.Pass{
		ResultOf: map[*analysis.Analyzer]interface{}{
			readme.Analyzer: jargonReadme,
			Analyzer:        nil,
		},
		Report: func(_ string, d analysis.Diagnostic) {
			invoked = true
			assert.Equal(t, "README.md contains developer jargon: (yarn)", d.Title)
		},
	}

	_, err := Analyzer.Run(pass)
	if err != nil {
		t.Fatal(err)
	}
	assert.True(t, invoked, "report should be called")
}
```

Considering this quite a common thing to do (checking if it has been invoked and running the corresponding function, there's also the same in `TestClean`), you can consider adding some helpers:

```go
// pkg/testcounter/testcounter.go
package testcounter

// FuncCallCounter is a type that implements a counter that can be used to count function invocations in tests.
// The counter can be incremented or reset to zero, but it can't be set to arbitrary values.
type FuncCallCounter struct {
	count int
}

// Inc increments the value of the counter by 1.
func (r *FuncCallCounter) Inc() {
	r.count++
}

// Value returns the current value of the counter.
func (r *FuncCallCounter) Value() int {
	return r.count
}

// Called returns true if the counter was incremented at least once.
func (r *FuncCallCounter) Called() bool {
	return r.count > 0
}

// CalledOnce returns true if the counter was incremented exactly once.
func (r *FuncCallCounter) CalledOnce() bool {
	return r.count == 1
}

// Reset resets the counter to zero.
func (r *FuncCallCounter) Reset() {
	r.count = 0
}
```

```go
// pkg/analysis/testinterceptor.go
package analysis

import "github.com/grafana/plugin-validator/pkg/testcounter"

// ReportCounter allows to create functions to be passed as "report" (func(string, Diagnostic)),
// that will also allow to count how many times the function was called.
type ReportCounter struct {
	testcounter.FuncCallCounter
}

// Function takes a function for a Report in a Pass, calls it and then increments the internal FuncCallCounter.
func (i *ReportCounter) Function(f func(string, Diagnostic)) func(string, Diagnostic) {
	return func(s string, d Diagnostic) {
		f(s, d)
		i.Inc()
	}
}
```

Then the `TestWithJargon` test becomes:

```go
func TestWithJargon(t *testing.T) {
	// This "intercept" variable allows us to track how many times a func(s string, d analysis.Diagnostic) is called.
	var intercept analysis.ReportCounter

	pass := &analysis.Pass{
		ResultOf: map[*analysis.Analyzer]interface{}{
			readme.Analyzer: jargonReadme,
			Analyzer:        nil,
		},
		// The func returned by "intercept.Function()" will call the passed func (the one with the "assert")
		// and also increment its internal counter, allowing us to assert on how many times the func was called.
		Report: intercept.Function(func(s string, d analysis.Diagnostic) {
			assert.Equal(t, "README.md contains developer jargon: (yarn)", d.Title)
		}),
	}

	_, err := Analyzer.Run(pass)
	if err != nil {
		t.Fatal(err)
	}
	
	// Assert on how many times the function was called.
	assert.True(t, intercept.CalledOnce(), "report should be called once")
}
```

Maybe a bit too overkill? If you have to do this a lot, I think it's worth checking it out (for a follow-up PR, if you plan to write tests for everything). It's also a good exercise on anonymous functions in go if you can figure out what's going on :)

Reply:
I appreciate the suggestion for the helpers. I'll take them in consideration for the next PR where I'll improve on the tests for all the passes. So far I pushed the check for invoked.

Reply:
:+1: Agree on keeping it in a separate PR if necessary